File: solutions/prolog/triangle/1/triangle.pl
triangle(Side1, Side2, Side3, Type) :- is_triangle(Side1, Side2, Side3),
    (((Type = "equilateral"; Type = "isosceles"), is_equilateral(Side1, Side2, Side3));
    (Type = "isosceles", is_isosceles(Side1, Side2, Side3));
    (Type = "scalene", is_scalene(Side1, Side2, Side3))).

is_equilateral(Side1, Side2, Side3) :- Side1 =:= Side2, Side1 =:= Side3.

is_isosceles(Side1, Side2, Side3) :-
    (Side1 =:= Side2, Side1 =\= Side3);
    (Side1 =:= Side3, Side1 =\= Side2);
    (Side2 =:= Side3, Side2 =\= Side1).

is_scalene(Side1, Side2, Side3) :- 
    (Side1 =\= Side2, Side1 =\= Side3, Side2 =\= Side3).
  
is_triangle(Side1, Side2, Side3) :-
    Side1 + Side2 >= Side3,
    Side2 + Side3 >= Side1,
    Side1 + Side3 >= Side2,
    Side1 =\= 0.

File: csharp/triangle/Triangle.cs
public static class Triangle
{
    public static bool IsScalene(double side1, double side2, double side3) =>
        IsValidTriangle(side1, side2, side3) && (side1 != side2 && side2 != side3 && side1 != side3);

    public static bool IsIsosceles(double side1, double side2, double side3) =>
        IsValidTriangle(side1, side2, side3) && (side1 == side2 || side2 == side3 || side1 == side3);

    public static bool IsEquilateral(double side1, double side2, double side3) =>
        IsValidTriangle(side1, side2, side3) && (side1 == side2 && side2 == side3);

    private static bool IsValidTriangle(double side1, double side2, double side3) =>
        side1 + side2 > side3 && side1 + side3 > side2 && side2 + side3 > side1;
}